test/typecheck/main.go

```diff
@@ -0,0 +1,347 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// do a fast type check of kubernetes code, for all platforms.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/build"
+	"go/parser"
+	"go/token"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/crypto/ssh/terminal"
+	// TODO(rmmh): remove this when golang/go#23712 is fixed, and the
+	// fix is the current minimum Go version to build Kubernetes.
+	"k8s.io/kubernetes/test/typecheck/srcimporter"
+	"k8s.io/kubernetes/third_party/forked/golang/go/types"
+)
+
+var (
+	verbose   = flag.Bool("verbose", false, "print more information")
+	cross     = flag.Bool("cross", true, "build for all platforms")
+	platforms = flag.String("platform", "", "comma-separated list of platforms to typecheck")
+	timings   = flag.Bool("time", false, "output times taken for each phase")
+	defuses   = flag.Bool("defuse", false, "output defs/uses")
+	serial    = flag.Bool("serial", false, "don't type check platforms in parallel")
+
+	isTerminal = terminal.IsTerminal(int(os.Stdout.Fd()))
+	logPrefix  = ""
+
+	// When processed in order, windows and darwin are early to make
+	// interesting OS-based errors happen earlier.
+	crossPlatforms = []string{
+		"linux/amd64", "windows/386",
+		"darwin/amd64", "linux/arm",
+		"linux/386", "windows/amd64",
+		"linux/arm64", "linux/ppc64le",
+		"linux/s390x", "darwin/386",
+	}
+)
+
+type analyzer struct {
+	fset      *token.FileSet // positions are relative to fset
+	conf      types.Config
+	ctx       build.Context
+	failed    bool
+	platform  string
+	donePaths map[string]interface{}
+}
+
+func newAnalyzer(platform string) *analyzer {
+	ctx := build.Default
```

Comment:
Consider adding an OpenFile func that uses a semaphore to limit the number of open files to something reasonable. It looks like srcimporter creates a goroutine per file unconstrained.

Reply:
In practice the most open files I've seen during an execution is ~20, so I'm not going to bother limiting it. Note that go vet etc use the same code paths without worrying about the number of open files.

Reply:
SGTM

Reply:
Argh, this can fail on OSX-- but there's no generic way to thread a semaphore through everywhere that it opens files.